src/signature.cc

```diff
@@ -0,0 +1,59 @@
+/* Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/signature.h"
+#include "utils/md5.h"
+
+using std::string;
+using ::istio::mixer::v1::CheckRequest;
+using ::istio::mixer::v1::CheckResponse;
+
+namespace istio {
+namespace mixer_client {
+namespace {
+const char kDelimiter[] = "\0";
+const int kDelimiterLength = 1;
+}  // namespace
+
+string GenerateAttributesSignature(const Attributes& attributes) {
+  MD5 hasher;
```

Comment:
std::string can be inside union, but it needs a manually allocation (in-place new) and deallocation to call constructor and destructor of std::string. There are similar codes in Envoy.